### DISCUSSION TITLE
通过 player manifest 获取视频信息

### DISCUSSION THREAD

**Discussion by WakelessSloth56** [the1812/Bilibili-Evolved#4330] [分享 / Share]
看到在议题 https://github.com/the1812/Bilibili-Evolved/issues/4303 中讨论的内容，搜索了一下 `manifest` 等关键词好像也没有人提到过相关内容，鉴于该议题已经关闭数天了，单独开一个讨论来讨论一下。

---

通过 `window.player` 的 `getManifest()` 方法可获取包括`aid`、`bvid`、`cid`、`seasonId`、`episodeId`等视频信息，例如：

- 一般视频 （video/BV1Xx411c7cH）
  ```json 
  {
	"p": 1,
	"aid": 120040,
	"cid": 201284,
	"pid": 2,
	"bvid": "BV1Xx411c7cH",
	"seasonId": null,
	"episodeId": null,
	"kind": 0,
	"screenKind": 0,
	"channelKind": 0,
	"seasonType": 0,
	"hasPrev": false,
	"hasNext": true,
	"hasPipFn": true,
	"hasFullscreenFn": true,
	"referrer": ""
  }
  ```
- 番剧 （bangumi/play/ep763748）
  ```json 
  {
	"p": 1,
	"aid": 743719840,
	"cid": 1203022611,
	"pid": 2,
	"bvid": null,
	"seasonId": 45503,
	"episodeId": 763748,
	"kind": 1,
	"screenKind": 0,
	"channelKind": 0,
	"seasonType": 1,
	"hasPrev": null,
	"hasNext": null,
	"hasPipFn": true,
	"hasFullscreenFn": true,
	"referrer": ""
  }
  ```
- 课程 （cheese/play/ep3248）
  - 课程页面没有 `window.player`，而是 `window.playerRaw`
  ```json 
  {
	"p": 2,
	"aid": 414784639,
	"cid": 240247575,
	"pid": 4,
	"bvid": null,
	"seasonId": 191,
	"episodeId": 3248,
	"kind": 2,
	"screenKind": 0,
	"channelKind": 0,
	"seasonType": 0,
	"hasPrev": true,
	"hasNext": true,
	"hasPipFn": true,
	"hasFullscreenFn": true,
	"referrer": ""
  }
  ```

**Comment by WakelessSloth56** [comment]
相较于[现在代码中用的 `__getUserParams()`](https://github.com/the1812/Bilibili-Evolved/blob/3e106af17c1ae4bd626e4c7d15791fcd18be51fc/src/components/video/player-adaptor/v4.ts#L15C47-L15C47)， `getManifest()` 不论是一般视频还是番剧等，获取到的字段都是相同的，而且 `hasPrev`、`hasNext` 等在一般视频或番剧下 `__getUserParams()` 获取到的中不存在，而在课程好像始终是 `null`。

**Comment by WakelessSloth56** [comment]
暂时我没有做更多的测试，不确定这个方法有没有特殊情况，或者是对不同版本的播放器有没有差别。

**Comment by the1812** [comment]
还有这种操作, 学到了

**Comment by timongh** [comment]
借着这个讨论我再提一嘴（总感觉以前在哪里提过）。B 站这个 window.player 可以和 window.nano 配合使用（视频、番剧、课程都有）

举几个我研究过的例子：
`player.on(nano.EventType.Player_Initialized, () => { /* ... */ }) `：监听视频开始播放
`player.once(nano.EventType.Player_Initialized, () => { /* ... */ }) `：监听视频开始播放，只执行一次
`player.off(nano.EventType.Player_Initialized, listener) `：停止监听
`player.setHandoff(nano.HandoffKind.Auto)`：开启自动连播
`player.setHandoff(nano.HandoffKind.Abort)`：关闭自动连播

最后，通过在 `BangumiPlayerAgent` 使用内置 api 可以顺手解决 issue: #2544

**Reply by WakelessSloth56** [reply]
之前确实有提到过：https://github.com/the1812/Bilibili-Evolved/pull/3381#issuecomment-1143756988
（不过我建议你可以就这个单独开一个讨论，在这里和这个讨论的题目关系不大，不利于后来人的检索。）

**Reply by timongh** [reply]
你说的对

新讨论：#4341
